app/PaymentDrivers/CheckoutCom/Utilities.php

```diff
@@ -39,4 +46,131 @@ public function convertToCheckoutAmount($amount, $currency)
+        $data = [
+            'payment_method' => $_payment->source['id'],
+            'payment_type' => PaymentType::parseCardType(strtolower($_payment->source['scheme'])),
+            'amount' => $this->payment_hash->data->value,
+        ];
+
+        $payment = $this->checkout->createPayment($data, \App\Models\Payment::STATUS_COMPLETED);
+
+        $this->payment_hash->payment_id = $payment->id;
+        $this->payment_hash->save();
+
+        $this->checkout->attachInvoices($payment, $this->payment_hash);
+
+        $payment->service()->updateInvoicePayment($this->payment_hash);
+
+        event(new PaymentWasCreated($payment, $payment->company, Ninja::eventVars()));
+
+        SystemLogger::dispatch(
+            ['response' => $_payment, 'data' => $data],
+            SystemLog::CATEGORY_GATEWAY_RESPONSE,
+            SystemLog::EVENT_GATEWAY_SUCCESS,
+            SystemLog::TYPE_CHECKOUT,
+            $this->checkout->client
+        );
+
+        return redirect()->route('client.payments.show', ['payment' => $this->checkout->encodePrimaryKey($payment->id)]);
+    }
+
+    public function processUnsuccessfulPayment(\Checkout\Models\Payments\Payment $_payment)
+    {
+        PaymentFailureMailer::dispatch(
+            $this->checkout->client,
+            $_payment,
+            $this->checkout->client->company,
+            $this->payment_hash->data->value
+        );
+
+        $message = [
+            'server_response' => $_payment,
+            'data' => $this->payment_hash->data,
+        ];
+
+        SystemLogger::dispatch(
+            $message,
+            SystemLog::CATEGORY_GATEWAY_RESPONSE,
+            SystemLog::EVENT_GATEWAY_FAILURE,
+            SystemLog::TYPE_CHECKOUT,
+            $this->checkout->client
+        );
+
+        return render('gateways.unsuccessful', [
+            'code' => $_payment->http_code,
+            'message' => $_payment->status,
+        ]);
+    }
+
+    private function processPendingPayment(\Checkout\Models\Payments\Payment $_payment)
+    {
+        $data = [
+            'payment_method' => $_payment->source['id'],
+            'payment_type' => PaymentType::parseCardType(strtolower($_payment->source['scheme'])),
+            'amount' => $this->payment_hash->data->value,
+        ];
+
+        $payment = $this->checkout->createPayment($data, \App\Models\Payment::STATUS_PENDING);
+
+        $this->payment_hash->payment_id = $payment->id;
+        $this->payment_hash->save();
+
+        $this->checkout->attachInvoices($payment, $this->payment_hash);
+
+        $payment->service()->updateInvoicePayment($this->payment_hash);
+
+        event(new PaymentWasCreated($payment, $payment->company, Ninja::eventVars()));
+
+        SystemLogger::dispatch(
+            ['response' => $_payment, 'data' => $data],
+            SystemLog::CATEGORY_GATEWAY_RESPONSE,
+            SystemLog::EVENT_GATEWAY_SUCCESS,
+            SystemLog::TYPE_CHECKOUT,
+            $this->checkout->client
+        );
```

Comment:
This shouldn't live here, this should be abstracted back to the BaseDriver.

Reply:
Fixed here: https://github.com/invoiceninja/invoiceninja/pull/4202/commits/f73ad4bef74a615fe195642422aafba5db05c43b